### ISSUE TITLE
gene predictions for second to last contig do not propagate to final gbk/gff 

### ISSUE THREAD

**Issue by VidJa** [nextgenusfs/funannotate#40]
After running the prediction step I get ~12k valid genemodels on 35 contigs in the predict_misc folder. However the predict_results folder only contains predictions for the first contig in the input fasta file.

**Comment by nextgenusfs** [comment]
Not seen this before either, can you post the log file?

**Comment by simas232** [comment]
I think I have an identical issue like @VidJa. The problem is that in the end of **funannotate predict** pipeline I get 1196 gene models, although there were 4860 gene models just before renaming the gene structures.

My system is CentOS 6.8. According to my investigation there are still 4860 gene models in gag2 folder and in mapping.ids, ncbi.cleaned.gff3 files. However, I have only 1196 gene structures in tbl2asn folder. I tried to re-run gag.py (last run in the pipeline) with the settings provided in funannotate-predict.log. The newly generated files in tbl2asn contain 4860 gene models, suggesting that there must be something wrong with tbl2asn function, which is the very last one in **funannotate predict** pipeline.

Thank you in advance for your support! I can attach the files from my pipeline, if requested.

**Comment by nextgenusfs** [comment]
@simas232 do you have the log file so I can see where you are going from 4860 to 1196?

**Comment by simas232** [comment]
[funannotate-predict.txt](https://github.com/nextgenusfs/funannotate/files/839960/funannotate-predict.txt)

**Comment by nextgenusfs** [comment]
Thanks.  Ok, how about the output of the following:
```
grep -c $'\tgene\t' kkmaI_results/predict_results/saccharomyces_cerevisiae_s288c.gff3
```

And how about this file please: 
```kkmaI_results/predict_misc/tlb2asn/genome.stats```

**Comment by simas232** [comment]
grep gives the number 1196

[genome.stats.txt](https://github.com/nextgenusfs/funannotate/files/839969/genome.stats.txt)

**Comment by nextgenusfs** [comment]
I don't think it is tbl2asn, but seems like the renaming script is perhaps failing?  What is the output of these two commands? I want to see if the renaming has worked or not?
```
grep $'\tgene\t' kkmaI_results/predict_misc/gag2/genome.gff | head 
```
and then this one (which is the renamed version)
```
grep $'\tgene\t' kkmaI_results/predict_misc/ncbi.cleaned.gff3 | head 
```
My assumption is that this command will yield 1196?
```
grep -c $'\tgene\t' kkmaI_results/predict_misc/ncbi.cleaned.gff3
```

**Comment by nextgenusfs** [comment]
So let me explain what is happening, maybe that will help narrow it down.  So after the second run through of GAG, output is in the gag2 folder.  The next step is to convert to preliminary genbank format and then parse the error report, and then go through and clean up the GFF file from gag2 directory, the output is then named ncbi.cleaned.gff.  So its plausible that the filtering isn't working, or also plausible that tbl2asn is an old version and the parser I wrote isn't doing what I want it to?  The current version of tbl2asn I'm running is 25.3, I can't tell what your version is from the log file.  Might be worth trying to update tbl2asn?

**Comment by simas232** [comment]
Sorry, @nextgenusfs, I didn't see that you asked for gag2/genome.stats. As I wrote before, gag2 folder contains 4860 gene structures. The file is attached to this message
[genome.stats.txt](https://github.com/nextgenusfs/funannotate/files/840488/genome.stats.txt)

**Comment by simas232** [comment]
The output for other commands are as the following:

`grep $'\tgene\t' kkmaI_results/predict_misc/gag2/genome.gff | head`
[output_before.txt](https://github.com/nextgenusfs/funannotate/files/840543/output_before.txt)

`grep $'\tgene\t' kkmaI_results/predict_misc/ncbi.cleaned.gff3 | head`
[output_after.txt](https://github.com/nextgenusfs/funannotate/files/840544/output_after.txt)

`grep -c $'\tgene\t' kkmaI_results/predict_misc/ncbi.cleaned.gff3`
returns the number 4851. This is strange, since I expected to see 4860

**Comment by simas232** [comment]
I have tbl2asn 25.3 as well.

When checking funannotate-predict.log once again, I noticed that tbl2asn is used twice in the pipeline, i.e. before and after the genes renaming. The file gag2/genome.gbf contains 4860 genes, suggesting that tbl2asn works fine and that there may be inconsistencies in the files, which we provide as input for the second tbl2asn run. According to my earlier investigation, the second run tbl2asn is the step where genes number is reduced from 4851 to 1196. We can therefore imply that this is not tbl2asn malfunctioning, but inconsistencies in the input files for the second run of tbl2asn. So taking it all together, something is wrong in genes renaming step.

I guess the answer lies in output_before.txt and output_after.txt, there must be something inconsistent in the latter file.

**Comment by nextgenusfs** [comment]
Yes, it also means the cleaning step is working fine, that's the number I see in the log file as well.  The renaming seems to be working based on the output files and that `ncbi.cleaned.gff3` is renamed in place, so it seems to have 4851 gene models going into GAG. 

What I really don't get is that the `ncbi.cleaned.gff3` file is what is being used to run GAG the final time (those results are in the `tbl2asn` folder).  But the `genome.stats` file from the `tbl2asn` folder says there are only 1196 genes, which is not what is in the `ncbi.cleaned.gff3` file.  Can you try to run this command manually and then investigate the results (i.e. looking at the `genome.stats` output?
```
gag.py -f kkmaI_results/predict_misc/genome.softmasked.fa \
-g kkmaI_results/predict_misc/ncbi.cleaned.gff3 \
-o kkmaI_results/predict_misc/tbl2asn_manual --fix_start_stop
```
Either that or if you want to post the soft masked genome file and the ncbi.cleaned.gff3 file I can check myself.

**Comment by simas232** [comment]
I ran gag.py with your suggested settings. I got the following output:
`grep -c $'gene' kkmaI_results/predict_misc/tbl2asn_manual/genome.gff`
4851
`grep -c $'gene' kkmaI_results/predict_misc/tbl2asn_manual/genome.tbl`
4851

tbl2asn_manual/genome.stats is also attached
[genome.stats.txt](https://github.com/nextgenusfs/funannotate/files/841726/genome.stats.txt)

**Comment by simas232** [comment]
I am thereby sending the files, required to run gag.py and then tbl2asn. Please try to run both gag.py and then tbl2asn. Maybe you will get 4851 genes in the end of the pipeline

[genome.softmasked.fa.txt](https://github.com/nextgenusfs/funannotate/files/841751/genome.softmasked.fa.txt)
[ncbi.cleaned.gff3.txt](https://github.com/nextgenusfs/funannotate/files/841753/ncbi.cleaned.gff3.txt)

**Comment by nextgenusfs** [comment]
Ok, will give it a try.  But doesn't make sense to me, as that is the same command that funannotate is running....  give me a few min

**Comment by nextgenusfs** [comment]
This all runs fine on my end, I end up with 4851 genes (4750 protein coding).  Gene models look reasonable and genbank file is formatted correctly.
```
-----------------------------------------
Saccharomyces cerevisae ???
-----------------------------------------
Scaffolds:	462
Total bp:	10.57 Mb
Largest Contig:	239,204 bp
Scaffold N50:	53,142 bp
GC Content:	40.04 %
Genes:		4,851
Protein Coding:	4,750
tRNA:		101
Avg gene len:	1,530 bp
```

So here's I guess what we can try.  Can you delete the following files from the 'predict_misc' folder and then re-run your original funannotate command (it will use the existing prediction data and re-run the formatting).
```
rm predict_misc/ncbi.cleaned.gff3
rm -r predict_misc/tbl2asn
```

**Comment by simas232** [comment]
I tried to run tbl2asn and now it seems that all the genes are now in place in tbl2asn file..

When I ran **funannot predict** again thereby having ncbi.cleaned.gff3 and tbl2asn deleted, the stuff became even more interesting. I got 2660 proteins in the predict_results and predict_misc/tbl2asn. As before, predict_misc/ncbi.cleaned.gff3 and predict_misc/gag2 contained 4800+ genes.

As much as I recall, the only change which I did to the system was installing maker2 instead maker3. 

Nevertheless, the stuff becomes too complicated for now, so for now it am running the fresh **funannotate predict** when using maker2 gff as input. I will let you know about the results once the pipeline is complete.

**Comment by nextgenusfs** [comment]
Hmm.  Good idea on the fresh run.  But again the predict_results/.gff3 file is actually copied from the tbl2asn gff output file.  The other place to look is in the tbl2asn output folder if the `genome.invalid.gff` and/or `genome.ignored.gff` - both of which should be empty.

Note, you will get different results if using Maker2 GFF as input to funanntoate, this is largely due to how Maker2 masks repeats versus how I do it in funanntoate.  For many fungi (probably not yeast), the repeat masking in Maker2 is too aggressive and you end up with a lower number of gene models than you should.  The renaming part of funannotate uses two Maker scripts, but the are packaged with funannotate thus that shouldn't be a conflict.

**Comment by MichaelFokinNZ** [comment]
genome.invalid.gff has _scaffold_4	EVM	gene_  (the place where annotation presumably stopped)
genome.ignored.gff is empty

**Comment by nextgenusfs** [comment]
This has bee fixed in v0.6.1 (I hope....)
